packagedb/find_source_repo.py

```diff
@@ -0,0 +1,214 @@
+from urllib.parse import urlparse
+
+import requests
+from packageurl import PackageURL
+from packageurl.contrib.purl2url import purl2url
+
+from packagedb.models import Package
+
+logger = logging.getLogger(__name__)
+
+
+def get_source_repo(package: Package) -> PackageURL:
+    """
+    Return the PackageURL of the source repository of a Package
+    or None if not found
+    """
+    repo_url = get_repo_url(package)
+    if not repo_url:
+        return
+    source_purl = convert_repo_url_to_purl(repo_url)
+    if not source_purl:
+        return
+    tag_and_commit = find_package_version_tag_and_commit(
+        version=package.version, source_purl=source_purl
+    )
+    if not tag_and_commit:
+        # TODO: determine what we do if do not have tag and commit
+        return
+    tag, commit = tag_and_commit
+    # if download_url can't be converted from a purl, use it as a qualifier
+    return PackageURL(
+        type=source_purl.type,
+        namespace=source_purl.namespace,
+        name=source_purl.name,
+        version=tag,
+        qualifiers={"commit": commit},
+    )
+
+
+def get_repo_url(package: Package) -> str:
+    """
+    Return the URL of the source repository of a package
+    """
+    source_url = get_source_url_from_package_data_and_resources(package=package)
+    if source_url:
+        return source_url
+    # We didn't find a source URL in the package data or resources
+    # Try all the package versions starting with the latest
+    # TODO: Use univers to sort versions
+    # TODO: Also consider using dates https://github.com/nexB/purldb/issues/136
+    for version_package in package.get_all_versions().order_by("-version"):
+        source_url = get_source_url_from_package_data_and_resources(
+            package=version_package
+        )
+        if source_url:
+            return source_url
+
+
+def get_source_url_from_package_data_and_resources(package: Package) -> str:
+    """
+    Return the URL of the source repository of a package
+    or None if not found
+    """
+    source_url = get_url_from_package_data(package)
+    if source_url:
+        return source_url
+    source_url = get_url_from_package_resources(package)
+    if source_url:
+        return source_url
+
+
+def convert_repo_url_to_purl(source_url):
+    """
+    Convert a source URL to a purl
+    """
+    url_hints = [
+        "github",
+        "gitlab",
+        "bitbucket",
+    ]
+    if not source_url:
+        return
+    # remove .git from the end of the URL
+    if source_url.endswith(".git"):
+        source_url = source_url[:-4]
+    if source_url.startswith("git://"):
+        # remove git:// from the beginning of the URL
+        source_url = source_url[6:]
+        if ":" in source_url:
+            # convert : to /
+            source_url = source_url.replace(":", "/")
+        source_url = f"https://{source_url}"
+    urlparse_result = urlparse(source_url)
+    if not len(urlparse_result.path.split("/")) > 2:
```

Comment:
`urlparse_result.path.split("/")` can be assigned to a variable, as it's repeated multiple times here